Source/WebCore/html/FormAssociatedCustomElement.h

```diff
@@ -0,0 +1,106 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "CustomElementFormValue.h"
+#include "HTMLElement.h"
+#include "ValidatedFormListedElement.h"
+#include "ValidityStateFlags.h"
+#include <JavaScriptCore/JSGlobalObject.h>
+
+namespace WebCore {
+
+class FormAssociatedCustomElement final : public ValidatedFormListedElement {
+    WTF_MAKE_NONCOPYABLE(FormAssociatedCustomElement);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static Ref<FormAssociatedCustomElement> create(HTMLElement&);
+
+    FormAssociatedCustomElement(HTMLElement&);
+    virtual ~FormAssociatedCustomElement();
+
+    bool isFormListedElement() const final { return true; }
+    bool isValidatedFormListedElement() const final { return true; }
+    bool isFormControlElement() const final { return false; }
+
+    FormAssociatedElement* asFormAssociatedElement() final { return this; }
+    FormListedElement* asFormListedElement() final { return this; }
+    ValidatedFormListedElement* asValidatedFormListedElement() final { return this; }
+
+    void reset() final;
+    bool isEnumeratable() const final;
+
+    void setFormValue(std::optional<CustomElementFormValue>&& submissionValue, std::optional<CustomElementFormValue>&& state);
+    ExceptionOr<void> setValidity(ValidityStateFlags, String&& message, HTMLElement* anchor);
+    String validationMessage() const final;
+
+    void formWillBeDestroyed() final;
+    void finishParsingChildren();
+
+    bool computeValidity() const final;
+    bool appendFormData(DOMFormData&) final;
+
+    void willUpgrade();
+    void didUpgrade();
+
+    const AtomString& formControlType() const final;
+    bool shouldSaveAndRestoreFormControlState() const final;
+    FormControlState saveFormControlState() const final;
+    void restoreFormControlState(const FormControlState&) final;
+
+protected:
+    void disabledStateChanged() final;
+
+private:
+    void refFormAssociatedElement() const final { m_element->ref(); }
+    void derefFormAssociatedElement() const final { m_element->deref(); }
+
+    HTMLElement& asHTMLElement() final { return *m_element.get(); }
+    const HTMLElement& asHTMLElement() const final { return *m_element.get(); }
+
+    HTMLElement* anchorElement() final;
```

Comment:
Naming this function anchorElement is quite confusing given we have HTMLAnchorElement.
Why not call this validationAnchorElement instead?
Since it implements https://html.spec.whatwg.org/#face-validation-anchor, right?

Reply:
Renamed, thanks, nice.